jenkins/aws-janitor/main.go

```diff
@@ -0,0 +1,786 @@
+		for _, vol := range page.Volumes {
+			v := &volume{Account: acct, Region: region, ID: *vol.VolumeId}
+			if set.Mark(v) {
+				glog.Warningf("%s: deleting %T: %v", v.ARN(), vol, vol)
+				toDelete = append(toDelete, v)
+			}
+		}
+		return true
+	}); err != nil {
+		return err
+	}
+	for _, vol := range toDelete {
+		_, err := svc.DeleteVolume(&ec2.DeleteVolumeInput{VolumeId: aws.String(vol.ID)})
+		if err != nil {
+			glog.Warningf("%v: delete failed: %v", vol.ARN(), err)
+		}
+	}
+	return nil
+}
+
+type volume struct {
+	Account string
+	Region  string
+	ID      string
+}
+
+func (vol volume) ARN() string {
+	return fmt.Sprintf("arn:aws:ec2:%s:%s:volume/%s", vol.Region, vol.Account, vol.ID)
+}
+
+// ARNs (used for uniquifying within our previous mark file)
+
+type arn struct {
+	partition    string
+	service      string
+	region       string
+	account      string
+	resourceType string
+	resource     string
+}
+
+func parseARN(s string) (*arn, error) {
+	pieces := strings.Split(s, ":")
+	if len(pieces) != 6 || pieces[0] != "arn" || pieces[1] != "aws" {
+		return nil, fmt.Errorf("Invalid AWS ARN: %v", s)
+	}
+	var resourceType string
+	var resource string
+	res := strings.SplitN(pieces[5], "/", 2)
+	if len(res) == 1 {
+		resource = res[0]
+	} else {
+		resourceType = res[0]
+		resource = res[1]
+	}
+	return &arn{
+		partition:    pieces[1],
+		service:      pieces[2],
+		region:       pieces[3],
+		account:      pieces[4],
+		resourceType: resourceType,
+		resource:     resource,
+	}, nil
+}
+
+func getAccount(sess *session.Session, region string) (string, error) {
+	svc := iam.New(sess, &aws.Config{Region: aws.String(region)})
+	resp, err := svc.GetUser(nil)
+	if err != nil {
+		return "", err
+	}
+	arn, err := parseARN(*resp.User.Arn)
+	if err != nil {
+		return "", err
+	}
+	return arn.account, nil
+}
+
+type s3path struct {
+	region string
+	bucket string
+	key    string
+}
+
+func getS3Path(sess *session.Session, s string) (*s3path, error) {
+	url, err := url.Parse(s)
+	if err != nil {
+		return nil, err
+	}
+	if url.Scheme != "s3" {
+		return nil, fmt.Errorf("Scheme %q != 's3'", url.Scheme)
+	}
+	svc := s3.New(sess, &aws.Config{Region: aws.String("us-east-1")})
+	resp, err := svc.GetBucketLocation(&s3.GetBucketLocationInput{Bucket: aws.String(url.Host)})
+	if err != nil {
+		return nil, err
+	}
+	region := "us-east-1"
+	if resp.LocationConstraint != nil {
+		region = *resp.LocationConstraint
+	}
+	return &s3path{region: region, bucket: url.Host, key: url.Path}, nil
+}
+
+func main() {
+	flag.Lookup("logtostderr").Value.Set("true")
+	flag.Parse()
+	regions := strings.Split(*regionsRaw, ",")
+
+	// Retry aggressively (with default back-off). If the account is
+	// in a really bad state, we may be contending with API rate
+	// limiting and fighting against the very resources we're trying
+	// to delete.
+	sess := session.Must(session.NewSessionWithOptions(session.Options{Config: aws.Config{MaxRetries: aws.Int(100)}}))
```

Comment:
The gotcha here is that if you've missed a dependency, you're going to hang for a long time waiting for the retry.

You may want to add something like https://github.com/kubernetes/kops/blob/fa42b60e9f61e1e6f25ea19d453ac992fc6f7f74/upup/pkg/fi/cloudup/awsup/logging_retryer.go